server/api/src/rs_genquery2.cpp

```diff
@@ -0,0 +1,131 @@
+namespace
+{
+    using log_api = irods::experimental::log::api;
+} // anonymous namespace
+
+auto rs_genquery2(RsComm* _comm, GenQuery2Input* _input, char** _output) -> int
+{
+    if (!_input || !_input->query_string || !_output) {
+        log_api::error("Invalid input: received nullptr for message pointer and/or response pointer.");
+        return SYS_INVALID_INPUT_PARAM;
+    }
+
+    // Redirect to the catalog service provider based on the user-provided zone.
+    // This allows clients to query federated zones.
+    //
+    // If the client did not provide a zone, getAndConnRcatHost() will operate as if the
+    // client provided the local zone's name.
+    if (_input->zone) {
+        log_api::trace("{}: Received: query_string=[{}], zone=[{}]", __func__, _input->query_string, _input->zone);
+    }
+    else {
+        log_api::trace("{}: Received: query_string=[{}], zone=[nullptr]", __func__, _input->query_string);
+    }
+
+    rodsServerHost* host_info{};
+
+    if (const auto ec = getAndConnRcatHost(_comm, PRIMARY_RCAT, _input->zone, &host_info); ec < 0) {
+        log_api::error("{}: Could not connect to remote zone [{}].", __func__, _input->zone);
+        return ec;
+    }
+
+    // Return an error if the host information does not point to the zone of interest.
+    // getAndConnRcatHost() returns the local zone if the target zone does not exist. We must catch
+    // this situation to avoid querying the wrong catalog.
+    if (host_info && _input->zone) {
+        const std::string_view resolved_zone = static_cast<zoneInfo*>(host_info->zoneInfo)->zoneName;
+
+        if (resolved_zone != _input->zone) {
+            log_api::error("{}: Could not find zone [{}].", __func__, _input->zone);
+            return SYS_INVALID_ZONE_NAME;
+        }
+    }
+
+    if (host_info->localFlag != LOCAL_HOST) {
+        log_api::trace("{}: Redirecting request to remote zone [{}].", __func__, _input->zone);
+        return rc_genquery2(host_info->conn, _input, _output);
+    }
+
+    //
+    // At this point, we assume we're connected to the catalog service provider.
+    //
+
+    try {
+        // FIXME "gq" is pre-defined in the Bison grammar.
+        // This needs to be removed or defined in a way that's internal/private to the grammar.
+        gq::options opts;
+
+        {
+            // Get the database type string from server_config.json.
+            const auto handle = irods::server_properties::instance().map();
+            const auto& config = handle.get_json();
+            const auto& db = config.at(nlohmann::json::json_pointer{"/plugin_configuration/database"});
+            opts.database = std::begin(db).key();
+        }
+
+        opts.username = _comm->clientUser.userName; // TODO Handle remote users?
+        opts.admin_mode = irods::is_privileged_client(*_comm); // TODO Should probably be a choice for the admin.
+        //opts.default_number_of_rows = 8; // TODO Can be pulled from R_GRID_CONFIGURATION on server startup.
```

Comment:
Any thoughts on the TODOs?

Reply:
those sound correct.  default of local zone.  default of 256.  option for admin to be non-admin.

Reply:
The TODO for L97 likely won't happen for this PR since that's actually new in iRODS.

If I did implement it, I was leaning towards `-m`, `--downgrade-privileges`, `--as-rodsuser`, or `--non-admin` as meaning run as a non-admin. Haven't given the naming much thought, yet.

Reply:
are you giving it a rodsusername to run 'as'?   or just... hmm...

we don't have this concept anywhere else I don't think.

we have `-M` which allows an admin to *use* their powers, but nothing that explicitly does NOT use their powers.

Reply:
> are you giving it a rodsusername to run 'as'? 

That wasn't my intention. the `-m` option would cause the parser to include permission checks.

Accepting a username sounds interesting. Maybe rodsadmins are allowed to run GenQuery2 as different users.

So ...
- `-m` - Run as admin but include permission checks.
- `-m <username>` - Run as `<username>`. Permission checks are based on `<username>`'s user type.

> we don't have this concept anywhere else I don't think.

Correct. It could be useful for debugging/testing purposes.

Reply:
This definitely requires more thinking/discussion.   Please create a new issue for this `permission checks` mode for genquery2.

Reply:
Created #7615.